ledger/src/bank_transaction_executor.rs

```diff
@@ -0,0 +1,547 @@
+    pub fn join(self) -> thread::Result<()> {
+        drop(self.sender);
+        for t in self.threads {
+            t.join()?;
+        }
+        Ok(())
+    }
+
+    fn start_execution_threads(
+        receiver: TransactionExecutionReceiver,
+        num_executors: usize,
+    ) -> Vec<JoinHandle<()>> {
+        (0..num_executors)
+            .map(|idx| {
+                let receiver = receiver.clone();
+
+                Builder::new()
+                    .name(format!("solBankTransactionExecutor-{}", idx))
+                    .spawn(move || Self::transaction_execution_thread(receiver))
+                    .unwrap()
+            })
+            .collect()
+    }
+
+    fn transaction_execution_thread(receiver: TransactionExecutionReceiver) {
+        while let Ok((
+            response_sender,
+            BankTransactionExecutionRequest {
+                bank,
+                tx,
+                transaction_status_sender,
+                replay_vote_sender,
+                cost_capacity_meter,
+            },
+        )) = receiver.recv()
+        {
+            let signature = *tx.signature();
+
+            let txs = vec![tx];
+            let mut batch = TransactionBatch::new(vec![Ok(())], &bank, Cow::Owned(txs));
+            batch.set_needs_unlock(false);
+
+            let mut timings = ExecuteTimings::default();
+            let execution_result = execute_batch(
+                &batch,
+                &bank,
+                transaction_status_sender.as_ref(),
+                replay_vote_sender.as_ref(),
+                &mut timings,
+                cost_capacity_meter,
+            );
+
+            if response_sender
+                .send(BankTransactionExecutionResponse {
+                    result: execution_result,
+                    timings,
+                    signature,
+                })
+                .is_err()
+            {
+                warn!("error sending back results");
+            }
+        }
+    }
+}
+
+fn execute_batch(
+    batch: &TransactionBatch,
+    bank: &Arc<Bank>,
+    transaction_status_sender: Option<&TransactionStatusSender>,
+    replay_vote_sender: Option<&ReplayVoteSender>,
+    timings: &mut ExecuteTimings,
+    cost_capacity_meter: Arc<RwLock<BlockCostCapacityMeter>>,
+) -> transaction::Result<()> {
+    let record_token_balances = transaction_status_sender.is_some();
+
+    let mut mint_decimals: HashMap<Pubkey, u8> = HashMap::new();
+
+    let pre_token_balances = if record_token_balances {
+        collect_token_balances(bank, batch, &mut mint_decimals, None)
+    } else {
+        vec![]
+    };
+
+    let pre_process_units: u64 = aggregate_total_execution_units(timings);
+
+    let (tx_results, balances) = batch.bank().load_execute_and_commit_transactions(
+        batch,
+        MAX_PROCESSING_AGE,
+        transaction_status_sender.is_some(),
+        transaction_status_sender.is_some(),
+        transaction_status_sender.is_some(),
```

Comment:
```suggestion
        record_token_balances,
        record_token_balances,
        record_token_balances,
```

Reply:
copied from blockstore_processor.rs, going to leave as is